### ISSUE TITLE
Mouse not working after last windows update

### ISSUE THREAD

**Issue by LordReven666** [yquake2/yquake2#257]
I had the same problem with quakespasm(SP for Q1) but they recently fixed this with a patch, ever since the last major windows update my mouse does not move in game.

**Comment by DanielGibson** [comment]
are you using 7.02 or 7.02a?
7.02a had critical fixes for Windows, you can get it at http://deponie.yamagi.org/quake2/windows/quake2-7.02a.zip

If that doesn't fix your problem, please try replacing the SDL2.dll with the one from http://libsdl.org/release/SDL2-2.0.7-win32-x86.zip - if there is a mouse problem on Windows, it's most probably an issue that would have to be fixed in SDL, maybe they have already done so before the latest release (2.0.7 is from October 22nd, and thus newer than the version we're bundling).

If it still doesn't work then, you'll need to tell us more about your system (which version of windows are you using, do you know which specific update broke the mouse input etc?)

**Comment by LordReven666** [comment]
Just tried both of the first suggestions and still not working.

I have a 64-bit Windows 10 Home edition PC with Intel i7 6700, Geforce 980, 16gb RAM, SoundBlaster Zx, and my mouse is an Mionix Avior(recently updated the firmware also) 

I believe the name of the update was the "Windows 10 Fall Creators update".

**Comment by ericwa** [comment]
AFAIK, we didn't fix anything in Quakespasm, just changed the default exe to be the SDL2 one.

My understanding is, SDL1.2's mouse input stopped working as of the Windows 10 Fall Creators update. SDL2's still worked (it uses the Windows raw input API by default), except if you set the `SDL_HINT_MOUSE_RELATIVE_MODE_WARP` hint to `1` (not the default).

This is the related SDL bug: https://bugzilla.libsdl.org/show_bug.cgi?id=3931

I haven't tried yquake2 but I can try reproducing this on Win 10 Fall Creators.

**Comment by ericwa** [comment]
I couldn't reproduce any mouse glitches with http://deponie.yamagi.org/quake2/windows/quake2-7.02a.zip
Tried both the included SDL2.dll and http://libsdl.org/release/SDL2-2.0.7-win32-x86.zip

This is Windows 10 Fall Creators Update (1709 build 16299.64)

btw, what exactly happens for you when the mouse input is broken @LordReven666 ?

**Comment by LordReven666** [comment]
Interesting, well when I am in game i cannot move my mouse, it is as if its stuck or something. Works fine in every other game I have now, except for yquake2.

**Comment by LordReven666** [comment]
I thought it was weird because I was having the same problems with quakespasm also before its update

**Comment by ericwa** [comment]
Hm - this might be unrelated then to the SDL bug I linked to ( https://bugzilla.libsdl.org/show_bug.cgi?id=3931 ). The symptom of that one is random jerks in the mouse motion (movement events when you don't move the mouse) as well as what feels like mouse events getting dropped.

**Comment by DanielGibson** [comment]
Oh, and one more idea: rename quake2.exe to yquake2.exe in case windows does some weird shit for the original q2 client based on the application name (unlikely, should have happened to Eric as well then, but might be worth a try)

**Comment by ericwa** [comment]
Oh - wow, never thought of that. I don't have any 3rd party security software.

It still is a bit confusing, I wonder why Quakespasm (and other games) work for @LordReven666 ?

I wonder if it would be worth trying the SDL2.dll from Quakespasm, overwriting the yquake2 one. (The SDL2.dll included in Quakespasm 0.93.0 is compiled from 2.0.7 sources + a few patches from bugzilla, nothing mouse related + disabling WASAPI audio. I.e. there's no good reason why it would fix anything, but this issue is still not quite making sense anyway..)

**Comment by LordReven666** [comment]
The only kind of security I have is windows defender which comes with windows 10, I believe it use to be called "Microsoft Security essentials".  

Moving the mouse moves the camera just a tiny bit in the direction the mouse is being moved, but then it just gets stuck and jerky. So I can't turn around or look anywhere at all. Here is a video of the problem: https://www.youtube.com/watch?v=Yl9QMMNuTSk I just streamed this 

Also, I just recently downloaded Quake2Pro and it works really well, and vanilla works fine, so windows has a problem with yquake2 it seems

**Comment by DanielGibson** [comment]
Ok, this looks very much like SDL bug 3931, which /should/ only happen if *either* that hint for mouse warping is set *or* for some reason enabling raw input fails and SDL falls back to mouse warping.
Not sure why raw input should fail, I doubt that Microsofts own Security thing breaks this.. and I have no idea at all why it works with Quakespasm but not YQ2..

Can you try renaming quake2.exe to yquake2.exe and using the SDL2.dll from Quakespasm?

**Comment by LordReven666** [comment]
now it just gives me this 
![capture](https://user-images.githubusercontent.com/33853428/33163194-1675196a-cfe2-11e7-99e3-d5dca64e4c41.PNG)

**Comment by DanielGibson** [comment]
ok, and if you just rename the executable, but keep our SDL2.dll (or the one downloaded from libsdl.org)?

**Comment by DanielGibson** [comment]
or try the SDL2.dll of the **32bit** version of Quakespasm - our windows binaries are 32bit, not 64bit

(while at it, you could also try if the 32bit version of Quakespasm works for you as well)

**Comment by DanielGibson** [comment]
So has this problem magically resolved itself, or..?

**Comment by LordReven666** [comment]
sorry have been busy lately

using the 32bit SDL2.dll did nothing, I still get the same problem, and yes Quakespasm 32bit works just fine for me

**Comment by DanielGibson** [comment]
Ok, can you try https://deponie.yamagi.org/quake2/misc/quake2_without_grab.zip ?

**Comment by LordReven666** [comment]
problem still exists

**Comment by DanielGibson** [comment]
Weird, I don't have many ideas anymore, especially as QS seems to work..

Are you playing in fullscreen or windowed mode (in both QS and YQ2)?

**Comment by LordReven666** [comment]
fullscreen in both

**Comment by DanielGibson** [comment]
Ok, can you try windowed mode in both? ("fullscreen: no" in our video menu)

And after that, can you try both "fullscreen: keep resolution" and "fullscreen: switch resolution" in YQ2?

Thanks!

**Comment by ericwa** [comment]
I guess the question is just why the Raw Input is failing for @LordReven666 , since everyone can reproduce that warping (CursorPos / SetCursorPos based) mouse input is broken on Win 10 Fall Creators.

I wonder if it would be worth compiling a custom sdl2.dll with logging added here:
https://hg.libsdl.org/SDL/file/93edd752e966/src/video/windows/SDL_windowsmouse.c#l36
and in particular, call GetLastError if the RegisterRawInputDevices returns FALSE?

**Comment by DanielGibson** [comment]
I'd still like windowed and the the two fullscreen modes tested, because that is the only further difference between YQ2 and QS I could imagine (the build linked above removed a call to SDL_SetWindowGrab() after setting relative mouse mode - this shouldn't be needed and QS doesn't do it, but apparently it didn't cause the problem either).
We use SDL_WINDOW_FULLSCREEN_DESKTOP by default, maybe QS uses SDL_WINDOW_FULLSCREEN and this makes a difference? In that case at least windowed mode should behave the same and at least one of the fullscreen modes in YQ2 should behave like QS.

Checking GetLastError() in SDL's ToggleRawInput() is a great idea; I'll try to provide such a SDL2.dll if the different fullscreen/windowed modes don't make a difference.

**Comment by LordReven666** [comment]
I tried everything you said and the problem was still there. :/

**Comment by DanielGibson** [comment]
So QS even works in windowed mode?

**Comment by DanielGibson** [comment]
Can you try the SDL2.dll from [sdl2.zip](https://github.com/yquake2/yquake2/files/1527347/sdl2.zip)?
It should show a popup window with an error code if RegisterRawInputDevices() fails.

Please tell us that errorcode (or post a screenshot of the popup).

**Comment by DanielGibson** [comment]
Update: That SDL2.dll showed the popup window even if no error happened, please use this one instead:
[sdl2fixed.zip](https://github.com/yquake2/yquake2/files/1528173/sdl2fixed.zip)

**Comment by DanielGibson** [comment]
Thanks a lot for testing!
According to https://msdn.microsoft.com/de-de/library/windows/desktop/ms681382(v=vs.85).aspx `87` means `ERROR_INVALID_PARAMETER`.

To me that doesn't make much sense, no idea what's going on there :-/
I had the theory that maybe your mouse doesn't identify as mouse but touchpad or whatever, and thus 
`RAWINPUTDEVICE rawMouse = { 0x01, 0x02, 0, NULL }; /* Mouse: UsagePage = 1, Usage = 2 */` doesn't work on your system - but I tried with my mouse disconnected (i.e. no mouse at all on the system) and didn't get the error, so that can't be it.
Unfortunately https://msdn.microsoft.com/en-us/library/windows/desktop/ms645600(v=vs.85).aspx doesn't have any information on when `RegisterRawInputDevices()` fails.

Maybe your mouse driver does something stupid or it's some kind of security software (or maybe a virus, the line between those is blurry anyway) that blocks `RegisterRawInputDevices()`?

And the fact that QS works and YQ2 still doesn't make any sense - does 32bit QS start with the SDL2.dll I linked above? Do you get the error there?

**Comment by DanielGibson** [comment]
Please test this SDL2.dll: [sdl2patched.zip](https://github.com/yquake2/yquake2/files/1532720/sdl2patched.zip)

(it has the change I suggest in https://discourse.libsdl.org/t/win10-fall-creators-update-breaks-mouse-warping/23526/2)

**Comment by ronroniv** [comment]
I'm not sure if this is related, but I encountered a similar problem on **Windows 7** 64bit / 7.02a.

In my case, the problem happens when not in fullscreen mode (fullscreen: no). Initially mouse movement works just fine in-game. But if I open menu and get back to in-game again, mouse movement stops working.

With 'switch resolution' and 'keep resolution' setting there's no problem.

I tested the above SDL2.dll just in case but the problem is still there.

**Comment by iButane** [comment]
@ronroniv nah, I believe that's a completely different issue. I used to get that a lot. I think it's just windows being stupid. I don't recall exactly how to fix it, but it was something simple like pressing alt, crtl or tab (or a combination) to re-enable the mouse movement. It's probably related to that whole 'application name' thing-- some people can't use their Windows key or alt tab once quake2.exe starts because of it.  There is a simply registry fix for it though. I think once I applied this I never again got the whole 'mouse frozen' thing after alt tabbing.

**Comment by DanielGibson** [comment]
if that's the case, just try renaming "quake2.exe" to "yq2.exe"

**Comment by iButane** [comment]
@DanielGibson yeah, but a lot of people like playing with the Steam overlay and it'll only launch quake2.exe.  This registry entry appears to fix it without renaming the original file: 
`Windows Registry Editor Version 5.00 

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\AppCompatFlags]
"{b1899c0f-fdfd-42d0-b489-c254bdbb539d}"=dword:00000001`

**Comment by iButane** [comment]
Looking through the Steam forums I noticed a few other people have the OP's mouse issue:

http://steamcommunity.com/app/2320/discussions/0/1480982338959928233/
http://steamcommunity.com/app/2320/discussions/0/2381701715726438917/

**Comment by LordReven666** [comment]
DanielGibson, that did it! Works great now, thank you so much! This is by far the best Q2 source port.

**Comment by DanielGibson** [comment]
Great!
We released 7.10 today and it includes (since Yamagi updated the zip ~2.5h ago) this patched SDL2.dll

At first we didn't want to add it as we don't know if it maybe breaks other things, but as more people seem to be affected by the bug (as can be seen in iButane's links) we decided to include it anyway.

This still is a workaround though and needs a proper fix in SDL

**Comment by ronroniv** [comment]
@iButane Thank you, it fixed my problem.

Also good to see that OP's problem has solved too!

**Comment by DanielGibson** [comment]
I've reported the issue to Microsoft, via their Feedback Hub App in Win10.
If you're running Win10, the following link should open the issue in the Feedback Hub so you can vote for it: https://aka.ms/W2icb9

Even though I've implemented a workaround in SDL2 (it was already merged by the SDL developers) it would be great if you could vote for the issue so Microsoft hopefully fixes it - after all it most probably affects lots of (especially older) games that can't be fixed as easily as SDL or YQ2.

**Comment by Novum** [comment]
I still have issues with the mouse stopping to work even with 7.10. After I access the console it's not working anymore.

**Comment by DanielGibson** [comment]
So this executable-specific breaking input comes from nvidia, not Windows? (If so, why do they mess with input, can't they stick to the rendering side of things?)

I guess we'll just rename the executable to yquake2.exe or something, and make quake2.exe a dummy that just starts the real one (to make sure the game can still be started from steam and that people updating their yq2 install don't accidentally start the old  executable after the update).

**Comment by DanielGibson** [comment]
I implemented the "just rename the executable to yquake2.exe or something, and make quake2.exe a dummy that just starts the real one", see #266 

Yamagi and me couldn't reproduce the bug, possibly because we have no Windows machine with a nvidia GPU, so it'd be nice if someone who is affected by the "mouse input doesn't work after openening and closing the console" bug (@Novum?) could test this :)
I attached the wrapper quake2.exe, it expects the real one to be named "yquake2.exe".

Thanks!

[quake2-wrapper.zip](https://github.com/yquake2/yquake2/files/1565323/quake2-wrapper.zip)

**Comment by ronroniv** [comment]
I'm not sure if it's related with Nvidia. I did run the game with both old integrated Intel HD graphics card and Nvidia card, and the issue exists either way.

Interestingly it shows me an error dialog box (probably because I'm still using the modified SDL2.dll from sdl2patched.zip above?) when I got into in-game:

> Title: RegisterRawInputDevices() for mouse failed!
> errcode:
> OK

(No error code is shown.)

Anyway I tested the wrapper. It works well and the mouse issue does not occur, the same as the renaming workaround.
